Added new auth-related methods and events to mcp server

etraut-openai opened the pull request:
This PR adds the following:
* A getAuthStatus method on the mcp server. This returns the auth method currently in use (chatgpt or apikey) or none if the user is not authenticated. It also returns the "preferred auth method" which reflects the `preferred_auth_method` value in the config.
* A logout method on the mcp server. If called, it logs out the user and deletes the `auth.json` file — the same behavior in the cli's `/logout` command.
* An `authStatusChange` event notification that is sent when the auth status changes due to successful login or logout operations.
* Logic to pass command-line config overrides to the mcp server at startup time. This allows use cases like `codex mcp -c preferred_auth_method=apikey`.

etraut-openai committed:
Added `getAuthStatus` and `logout` methods and `authStatusChanged` event notification to the MCP server.

etraut-openai committed:
Fixed lint

bolinfest commented on codex-rs/mcp-server/src/main.rs:
```diff
@@ -1,9 +1,10 @@
 fn main() -> anyhow::Result<()> {
     arg0_dispatch_or_else(|codex_linux_sandbox_exe| async move {
-        run_main(codex_linux_sandbox_exe).await?;
+        run_main(codex_linux_sandbox_exe, CliConfigOverrides::default()).await?;
```
Does this work? I notice we do something slightly more complex in https://github.com/openai/codex/blob/main/codex-rs/exec/src/main.rs.

bolinfest commented on codex-rs/mcp-server/src/lib.rs:
```diff
@@ -77,10 +82,26 @@ pub async fn run_main(codex_linux_sandbox_exe: Option<PathBuf>) -> IoResult<()>
 
+    // Parse CLI overrides once so we can reuse them for operations
+    // like `get_auth_status` that need access to the effective config.
+    let cli_kv_overrides = match cli_config_overrides.parse_overrides() {
+        Ok(v) => v,
+        Err(e) => {
+            return Err(std::io::Error::new(
+                std::io::ErrorKind::InvalidInput,
+                format!("error parsing -c overrides: {e}"),
+            ));
+        }
+    };
+
```
FYI, you could do `let cli_kv_overrides = cli_config_overrides.parse_overrides().map_err(...)?;` if you wanted to avoid the `match` and try to be a bit more succinct.

bolinfest commented on codex-rs/mcp-server/src/lib.rs:
```diff
@@ -77,10 +82,26 @@ pub async fn run_main(codex_linux_sandbox_exe: Option<PathBuf>) -> IoResult<()>
+    let cli_kv_overrides = match cli_config_overrides.parse_overrides() {
+        Ok(v) => v,
+        Err(e) => {
+            return Err(std::io::Error::new(
```
Prefer importing stdlib types at the top, though admittedly something common like `Error` might need to be fully-qualified inline (but `ErrorKind` is less of an issue).

bolinfest commented on codex-rs/mcp-server/src/lib.rs:
```diff
@@ -77,10 +82,26 @@ pub async fn run_main(codex_linux_sandbox_exe: Option<PathBuf>) -> IoResult<()>
+        let mut processor = MessageProcessor::new(
+            outgoing_message_sender,
+            codex_linux_sandbox_exe,
+            cli_kv_overrides,
```
I think `cli_kv_overrides` is an awkward type to thread through. `MessageProcessor` should not have to know about `TomlValue`, for example. I would derive the `Config` from it as early as possible and pass that through instead.

bolinfest commented on codex-rs/mcp-server/src/codex_message_processor.rs:
```diff
@@ -212,6 +229,20 @@ impl CodexMessageProcessor {
+                        };
+                        let params = serde_json::to_value(&notification).ok();
+                        outgoing_clone
+                            .send_notification(OutgoingNotification {
```
With my recommended change to introduce `ServerNotification`, we should update `send_notification` to take `ServerNotification`, which will make this simpler.

etraut-openai commented on codex-rs/protocol/src/protocol.rs:
```diff
@@ -859,3 +859,28 @@ mod tests {
     }
 }
+
+#[derive(Clone, Debug, PartialEq, Copy, Eq, Serialize, Deserialize, TS)]
```
The challenge I had here was that this didn't have the TypeScript converter. I tried adding it, but it introduced new dependencies to the login crate (it currently has no knowledge of TS conversions), and it didn't seem prudent to add the dependency for just this one enum. What do you think?

etraut-openai committed:
Added ServerNotification enum per code review feedback.

etraut-openai committed:
Moved AuthMode definition into mcp_server and changed login crate to import it, rather than the other way around.

etraut-openai committed:
Code review feedback.

etraut-openai committed:
Code review feedback: load config and apply overrides earlier in call chain.

etraut-openai committed:
Fixed server method names.

etraut-openai committed:
Merge branch 'main' into auth_status_logout

etraut-openai commented on codex-rs/mcp-server/src/outgoing_message.rs:
```diff
@@ -190,6 +195,30 @@ pub(crate) struct OutgoingNotification {
+    fn into_outgoing_notification(self) -> OutgoingNotification {
+        // Prefix all Codex server notifications with the MCP event namespace.
+        let method = format!("codex/event/{}", self);
+        let params = match &self {
```
I'm not sure this is the best way to serialize the payload. I'd prefer to find a solution that doesn't involve a match statement here, but I'm hitting the limits of my Rust knowledge, and GPT-5 is not coming up with any better suggestions.

etraut-openai commented:
@bolinfest, ready for another round of review. I incorporated feedback from your last round:
1. Added `ServerNotification` enum and updated `send_notification` to accept it.
2. Reversed the dependency between the `protocol` and `logon` crates by moving `AuthMode` from the latter to the former.
3. Precompute Config with overrides and pass it to mcp server rather than threading config overrides down the call chain.

etraut-openai committed:
Code review feedback.

etraut-openai committed:
Simplified serialization of ServerNotification objects.

bolinfest commented on codex-rs/mcp-server/src/outgoing_message.rs:
```diff
@@ -121,6 +122,17 @@ impl OutgoingMessageSender {
     }
 
+    pub(crate) async fn send_server_notification(&self, notification: ServerNotification) {
+        let method = format!("codex/event/{}", notification);
```
I actually want to drop the `codex/event` prefix, so if we go to:

```
#[serde(tag = "method", rename_all = "camelCase")]
pub enum ServerNotification {
```

then I think this gets even simpler?

etraut-openai replied:
Deferring this to a future PR as per 1:1 discussion.

bolinfest commented on codex-rs/protocol/src/mcp_protocol.rs:
```diff
@@ -321,6 +349,34 @@ pub struct ApplyPatchApprovalResponse {
+#[strum(serialize_all = "snake_case")]
+pub enum ServerNotification {
+    /// Authentication status changed
+    AuthStatusChange(AuthStatusChangeNotification),
```
With my above recommendation, I think this becomes:

```rust
AuthStatusChange {
    params: AuthStatusChangeNotification
}
```

bolinfest commented on codex-rs/protocol/src/mcp_protocol.rs:
```diff
@@ -321,6 +349,34 @@ pub struct ApplyPatchApprovalResponse {
+#[derive(Serialize, Deserialize, Debug, Clone, PartialEq, TS, Display)]
+#[serde(tag = "type", content = "data", rename_all = "snake_case")]
+#[strum(serialize_all = "snake_case")]
+pub enum ServerNotification {
```
Can you please add this enum to:

https://github.com/openai/codex/blob/9193eb6b53173962b58886b2d5fb33fe282d9aca/codex-rs/protocol-ts/src/lib.rs#L18

bolinfest reviewed (approved):
Nice work, but please address comments before submitting!

etraut-openai committed:
Generate ts bindings for ServerNotification.

etraut-openai committed:
Merge branch 'main' into auth_status_logout
